perception/autoware_tensorrt_bevdet/src/gatherbev_plugin.cu

```diff
@@ -0,0 +1,308 @@
+{
+  memcpy(&m_, buffer, sizeof(m_));
+}
+
+GatherBEVPlugin::~GatherBEVPlugin()
+{
+}
+
+IPluginV2DynamicExt * GatherBEVPlugin::clone() const noexcept
+{
+  auto p = new GatherBEVPlugin(name_, &m_, sizeof(m_));
+  p->setPluginNamespace(namespace_.c_str());
+  return p;
+}
+
+int32_t GatherBEVPlugin::getNbOutputs() const noexcept
+{
+  return 1;
+}
+
+DataType GatherBEVPlugin::getOutputDataType(
+  int32_t index, DataType const * inputTypes, int32_t nbInputs) const noexcept
+{
+  return inputTypes[0];  // 与adj_feat一致
+}
+
+DimsExprs GatherBEVPlugin::getOutputDimensions(
+  int32_t outputIndex, const DimsExprs * inputs, int32_t nbInputs,
+  IExprBuilder & exprBuilder) noexcept
+{
+  // input[0] == adj_feat   b x 8 x 80 x 128 x 128
+  // input[1] == curr_feat  b x 80 x 128 x 128
+  // input[2] == flag       b x 1
+  // out[0]                 b x (8+1)*80 x 128 x 128
+  DimsExprs ret;
+  ret.nbDims = inputs[0].nbDims - 1;
+
+  IDimensionExpr const * n_feat =
+    exprBuilder.operation(DimensionOperation::kSUM, *inputs[0].d[1], *exprBuilder.constant(1));
+  ret.d[0] = inputs[0].d[0];
+  ret.d[1] = exprBuilder.operation(DimensionOperation::kPROD, *n_feat, *inputs[0].d[2]);
+  ret.d[2] = inputs[0].d[3];
+  ret.d[3] = inputs[0].d[4];
+
+  return ret;
+}
+
+bool GatherBEVPlugin::supportsFormatCombination(
+  int32_t pos, const PluginTensorDesc * inOut, int32_t nbInputs, int32_t nbOutputs) noexcept
+{
+  // adj_feat
+  if (pos == 0) {
+    return (inOut[pos].type == DataType::kFLOAT || inOut[pos].type == DataType::kHALF) &&
+           inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 1) {  // curr_feat
+    return inOut[0].type == inOut[1].type && inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 3) {  // out
+    return inOut[0].type == inOut[3].type && inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 2) {
+    return inOut[pos].type == DataType::kINT32 && inOut[pos].format == TensorFormat::kLINEAR;
+  }
+  return false;
+}
+
+void GatherBEVPlugin::configurePlugin(
+  const DynamicPluginTensorDesc * in, int32_t nbInputs, const DynamicPluginTensorDesc * out,
+  int32_t nbOutputs) noexcept
+{
+  return;
+}
+
+size_t GatherBEVPlugin::getWorkspaceSize(
+  const PluginTensorDesc * inputs, int32_t nbInputs, const PluginTensorDesc * outputs,
+  int32_t nbOutputs) const noexcept
+{
+  return 0;
+}
+
+int32_t GatherBEVPlugin::enqueue(
+  const PluginTensorDesc * inputDesc, const PluginTensorDesc * outputDesc,
+  const void * const * inputs, void * const * outputs, void * workspace,
+  cudaStream_t stream) noexcept
+{
+  // input[0] == adj_feat   b x 8 x 80 x 128 x 128
+  // input[1] == curr_feat  b x 80 x 128 x 128
+  // input[2] == flag       b x 1
+  // out[0]                 b x (8+1)*80 x 128 x 128
+
+  // int nthreads, // b * (adj_num + 1) * map_size
+  // int adj_num,
+  // int channel,
+  // int map_size,
+
+  // int flag = 0;
+  // CHECK_CUDA(cudaMemcpy(&flag, inputs[2], sizeof(int), cudaMemcpyDeviceToHost));
```

Comment:
The code currently copies the `flag` variable from the device to the host. This introduces unnecessary overhead.  It's better to perform the conditional logic directly on the device using the `flag` values without host-device transfers.  This might involve modifying the `copy_feat_kernel` to handle the conditional logic efficiently using CUDA intrinsics for better performance.